File: app/src/main/java/com/mert/bilgram/FavoritesActivity.java
package com.mert.bilgram;

import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

public class FavoritesActivity extends AppCompatActivity {

    ArrayList<String> userEmailFromFB = new ArrayList<>(), userImageFromFB = new ArrayList<>(), userPostDescFromFB = new ArrayList<>(), postIDsFromFB = new ArrayList<>();
    FavoritesRecyclerViewAdapter adapter = new FavoritesRecyclerViewAdapter(userEmailFromFB, userPostDescFromFB, userImageFromFB, this, this);
    RecyclerView recyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (AppCompatDelegate.getDefaultNightMode() == AppCompatDelegate.MODE_NIGHT_YES)
            setTheme(R.style.AppThemeDark);

        setContentView(R.layout.activity_favorites);
        setTitle("Liked Posts");

        recyclerView = findViewById(R.id.favRecyclerView);
        recyclerView.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        ItemClickSupport.addTo(recyclerView).setOnItemClickListener(new ItemClickSupport.OnItemClickListener() {
            @Override
            public void onItemClicked(RecyclerView recyclerView, int position, View v) {
                Intent intent = new Intent(getApplicationContext(), ViewImageActivity.class);
                intent.putExtra("downloadURL", userImageFromFB.get(position));
                startActivity(intent);
            }

            @Override
            public void onItemDoubleClicked(RecyclerView recyclerView, int position, View v) {

            }
        });

        try {

            FeedActivity.database = this.openOrCreateDatabase("SavedPosts", MODE_PRIVATE, null);
            FeedActivity.database.execSQL("CREATE TABLE IF NOT EXISTS posts (id VARCHAR, url VARCHAR, postDesc VARCHAR, useremail VARCHAR, time VARCHAR)");

            Cursor cursor = FeedActivity.database.rawQuery("SELECT * FROM posts ORDER BY time DESC", null);

            int urlIndex = cursor.getColumnIndex("url");
            int postDescIndex = cursor.getColumnIndex("postDesc");
            int userEmailIndex = cursor.getColumnIndex("useremail");
            int IdIndex = cursor.getColumnIndex("id");

            cursor.moveToFirst();

            while (cursor != null) {

                if (!postIDsFromFB.contains(cursor.getString(IdIndex)))
                    postIDsFromFB.add(cursor.getString(IdIndex));
                else {
                    cursor.moveToNext();
                    continue;
                }
                userImageFromFB.add(cursor.getString(urlIndex));
                userEmailFromFB.add(cursor.getString(userEmailIndex));
                userPostDescFromFB.add(cursor.getString(postDescIndex));
                adapter.notifyDataSetChanged();

                cursor.moveToNext();

            }

            cursor.close();


        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    public void unSave(final int i) {

        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Remove")
                .setMessage("Do you want to remove this post from your Favourites")
                .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i2) {
                        String id = postIDsFromFB.get(i);

                        FeedActivity.database.execSQL("Delete from posts where id='" + id + "'");

                        Toast.makeText(FavoritesActivity.this, "Post removed from Favourites", Toast.LENGTH_SHORT).show();

                        userEmailFromFB.remove(i);
                        userImageFromFB.remove(i);
                        postIDsFromFB.remove(i);
                        userPostDescFromFB.remove(i);

                        adapter.notifyDataSetChanged();
                    }
                }).setNegativeButton("No", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {

            }
        }).show();

    }

}

File: app/src/main/java/com/mert/bilgram/FavoritesActivity.kt
package com.mert.bilgram

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import java.util.*

class FavoritesActivity : AppCompatActivity() {

    private var userEmailFromFB = ArrayList<String>()
    private var userImageFromFB = ArrayList<String>()
    private var userPostDescFromFB = ArrayList<String>()
    private var postIDsFromFB = ArrayList<String>()
    private var adapter = FavoritesRecyclerViewAdapter(userEmailFromFB, userPostDescFromFB, userImageFromFB, this, this)
    private lateinit var recyclerView: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (AppCompatDelegate.getDefaultNightMode() == AppCompatDelegate.MODE_NIGHT_YES)
            setTheme(R.style.AppThemeDark)

        setContentView(R.layout.activity_favorites)

        title = getString(R.string.liked_posts)

        recyclerView = findViewById(R.id.favRecyclerView)
        recyclerView.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(this)

        ItemClickSupport.addTo(recyclerView).setOnItemClickListener(object : ItemClickSupport.OnItemClickListener {
            override fun onItemClicked(recyclerView: RecyclerView?, position: Int, v: View?) {
                startActivity(Intent(applicationContext, ViewImageActivity::class.java).putExtra("downloadURL", userImageFromFB[position]))
            }

            override fun onItemDoubleClicked(recyclerView: RecyclerView?, position: Int, v: View?) {}
        })

        try {
            FeedActivity.database = this.openOrCreateDatabase("SavedPosts", Context.MODE_PRIVATE, null)

            FeedActivity.database.execSQL("CREATE TABLE IF NOT EXISTS posts (id VARCHAR, url VARCHAR, postDesc VARCHAR, useremail VARCHAR, time VARCHAR)")

            val cursor = FeedActivity.database.rawQuery("SELECT * FROM posts ORDER BY time DESC", null)

            val urlIndex = cursor!!.getColumnIndex("url")

            val postDescIndex = cursor.getColumnIndex("postDesc")

            val userEmailIndex = cursor.getColumnIndex("useremail")

            val idIndex = cursor.getColumnIndex("id")

            cursor.moveToFirst()

            do {
                if (!postIDsFromFB.contains(cursor.getString(idIndex)))
                    postIDsFromFB.add(cursor.getString(idIndex))
                else
                    continue

                userImageFromFB.add(cursor.getString(urlIndex))
                userEmailFromFB.add(cursor.getString(userEmailIndex))
                userPostDescFromFB.add(cursor.getString(postDescIndex))
                adapter.notifyDataSetChanged()
            } while (cursor.moveToNext())

            cursor.close()

        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun unSave(i: Int, v: View) {
        AlertDialog.Builder(this).setTitle(getString(R.string.remove))
                .setMessage(getString(R.string.remove_from_favourites))
                .setPositiveButton(R.string.yes) { _, _ ->

                    FeedActivity.database.execSQL("Delete from posts where id='${postIDsFromFB[i]}'")
                    Snackbar.make(v, R.string.removed_from_favourites, Snackbar.LENGTH_SHORT).show()
                    userEmailFromFB.removeAt(i)
                    userImageFromFB.removeAt(i)
                    postIDsFromFB.removeAt(i)
                    userPostDescFromFB.removeAt(i)
                    adapter.notifyItemRemoved(i)

                }.setNegativeButton(R.string.no) { _, _ -> }.show()
    }
}